File: src/206.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        #昨天面试字节碰上了它的高级题，力扣25，我先来欺负下它的小弟
        #用递归做一下
        #十分钟，一次提交，时间击败-5.36%，空间击败-5.54%
        #昨天要是做这个题我就搞定了呀
        if head == None or head.next ==  None:
            return head
        #先把末尾给标记一下
        fakeHead  = ListNode(-1)
        temp = head
        while temp.next != None:
            temp = temp.next
        fakeHead.next  =  temp  #锁定结尾
        #开始翻转，递归
        self.re_reverseList(head, temp)
        #处理尾节点
        head.next = None
        return  fakeHead.next

    def re_reverseList(self, _head, _tail):
        if _head == _tail:
            #头尾重合，翻转终止
            return
        else:
            #否则，翻转
            prevNode = _head
            while prevNode.next != _tail:
                prevNode = prevNode.next
            #翻转
            _tail.next = prevNode
            #向下
            self.re_reverseList(_head, prevNode)
            return

File: src/206.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        //后天面试字节，冲呀
        //原地翻转
        //On时间复杂度
        //边界情况
        //十分钟，两次提交，时间击败-95.76%, 空间击败-99.44%
        if(head == nullptr || head->next ==  nullptr){
            return head;
        }
        ListNode * prev = nullptr;  //前一节点
        ListNode * p = head;    //当前翻转节点
        //先拿到尾巴
        while(p->next){
            p =  p->next;
        }
        ListNode * tail = p;
        p = head;
        while(prev != tail){
            ListNode * node = p->next;  //保存下一节点
            //翻转
            p->next = prev;
            //移动
            prev  = p;
            p = node;
        }

        return tail;
    }
};
